labgrid/util/ssh.py

```diff
@@ -0,0 +1,324 @@
+            return self._forwards[destination]
+        self._run_socket_command(
+            "forward", " -L {local}:{destination}".format(
+                local=local_port, destination=destination
+            )
+        )
+        self._forwards[destination] = local_port
+        return local_port
+
+    @check_connected
+    def remove_port_forward(self, remote_host, remote_port):
+        """cancel command"""
+        destination = "{}:{}".format(remote_host, remote_port)
+        local_port = self._forwards.pop(destination, None)
+
+        # pylint: disable=not-an-iterable
+        if local_port is None:
+            raise ForwardError("Forward does not exist")
+
+        self._run_socket_command(
+            "cancel", " -L {local}:localhost:{remote}".format(
+                local=local_port, remote=remote_port
+            )
+        )
+
+    def connect(self):
+        self._open_connection()
+        self._connected = True
+
+    @check_connected
+    def disconnect(self):
+        self._disconnect()
+
+    def isconnected(self):
+        return self._connected
+
+    def _check_master(self):
+        args = ["ssh", "-O", "check", "{}".format(self.host)]
+        check = subprocess.call(
+            args,
+            stdin=subprocess.DEVNULL,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL
+        )
+        if check == 0:
+            return ""
+
+        return self._start_own_master()
+
+    def _start_own_master(self):
+        """Starts a controlmaster connection in a temporary directory."""
+        control = os.path.join(self._tmpdir, 'control-{}'.format(self.host))
+        args = (
+            "ssh -n {} -x -o ConnectTimeout=30 -o ControlPersist=300 "
+            "-o UserKnownHostsFile=/dev/null "
+            "-o StrictHostKeyChecking=no -MN -S {} {}"
+        ).format(self._ssh_prefix, control, self.host).split(" ")
+
+        self.process = subprocess.Popen(
+            args,
+            stderr=subprocess.PIPE,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE
+        )
+
+        try:
+            if self.process.wait(timeout=30) is not 0:
+                raise ExecutionError(
+                    "failed to connect to {} with {} and {} [{}],[{}] ".format(
+                        self.host, args, self.process.wait(), self.process.stdout.readlines(), self.process.stderr.readlines()
+                    )
+                )
+        except subprocess.TimeoutExpired:
+            raise ExecutionError(
+                "failed to connect to {} with {} and {} [{}],[{}]".format(
+                    self.host, args, self.process.readlines(), self.process.readlines()
+                )
+            )
+
+        if not os.path.exists(control):
+            raise ExecutionError("no control socket to {}".format(self.host))
+
+        self._logger.debug('Connected to %s', self.host)
+
+        return control
+
+    def _disconnect(self):
```

Comment:
What happens with existing forwards in this case?

Reply:
AFAIR forwards are closed if the socket is closed.

Reply:
I changed it to explicitly close the forwards before closing the socket